enhancements/update/upgrade-blocker-operator.md

```diff
@@ -0,0 +1,128 @@
+---
+title: upgrade-blocker-operator
+authors:
+  - "@michaelguigno"
+reviewers:
+  - TBD
+approvers:
+  - TBD
+creation-date: 2021-02-03
+last-updated: 2021-02-03
+status: implementable
+---
+
+# Upgrade Blocker Operator
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+Add an extensible API and associated operator to block upgrades.
+
+## Motivation
+
+Sometimes clusters shouldn't be upgraded.  Currently, the only signal is
+ClusterOperator statuses monitored by the CVO.  Cluster administrators need
+to be able to optionally inform the cluster when upgrades aren't appropriate.
+
+### Goals
+
+* Define a new user-facing CRD and namespace to be consume by the new operator
+* Create a new operator to signal whether or not upgrades can commence
+* Define an override mechanism to allow upgrades anyway
+
+### Non-Goals
+
+* Limit the scope of what information might be used to block an upgrade
+* Changes to the CVO other than respecting the new cluster operator
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As a cluster administrator, I need a pluggable system to inform the CVO when
+upgrades/automatic upgrades should be prevented.  Data I use to make this
+determination may originate inside or outside of the cluster, depending on
+my organization's needs.
+
+### Risks and Mitigations
+
+#### Yet another operator
+
+We could build some of this logic into the CVO directly instead of another
+cluster operator.
+
+#### People will want to block things other than upgrades
+
+This is not addressed here.
+
+## Design Details
+
+### New CRD
+
+A new CRD is created.  One or more instances (CRs) are to be created in a
+system-specific namespace.  The CR should capture the who/what/why, as well
+as severity.
+
+The CRD should have a field in spec "preventUpgrades" with a boolean
+value to clearly indicate if a given CR is block an upgrade or not.
```

Comment:
nit: boolean will not address minor (4.y -> 4.y') vs. patch (4.y.z -> 4.y.z') bumps.  Probably better to use an enumerated type.

Reply:
This is an open question if we need to add such an interface.  Probably we should, and so we might need something like
"blockAllUpgrades" and "blockMinorUpgrades"  I dislike that we use the term 'minor' because they're IMO not really 'minor' upgrades, but the wording can be ironed out I think.

Reply:
All this enhancement was ARO request. We will be doing this without this (just by creating CO CRD for ARO operator) and managing states there.
What we could really use is 
`Upgradable`, and `UpgradableMinor` or something like this. 

So if this enhancement will get closed we would not get sad :)

Reply:
> I dislike that we use the term 'minor' because they're IMO not really 'minor' upgrades...

[SemVer spec][1]:

> Given a version number MAJOR.MINOR.PATCH...

Easier to not invent new, OpenShift-specific words for the different SemVer components.

> What we could really use is Upgradable, and UpgradableMinor or something like this.

When would you allow minor bumps but block patch bumps?  I can see cases for "allow everything", "block minor bumps, allow patch bumps", and "block everything".  You can handle that with a single knob and an enumerated list of values.  You can't handle that with two separate knobs, without also exposing the nonsensical "allow minor bumps, block patch bumps".

And yeah, we can probably close this now that ARO is going with an admission webhook in front of ClusterVersion, and revisit if/when we get another use-case where a generic operator would be useful.

[1]: https://semver.org/spec/v2.0.0.html

Reply:
ARO is only one user, and IMO, webhooks as a way to block upgrades is not the best design.  We should build an API that allows any user to block an upgrade if they choose.

User intent being translated into actions on the cluster, that's an API.  We use CRDs to build those APIs today.  We can use the existing ClusterOperator API (upgradeable=False) with a new user-facing API to create this functionality.